src/terrain/KM_Terrain.pas

```diff
@@ -585,31 +622,210 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
   end;
 
   //Did block diagonal property change? (hence xor) UpdateWalkConnect needs to know
-  DiagonalChanged := gMapElements[Land[Y,X].Obj].DiagonalBlocked xor gMapElements[aObject].DiagonalBlocked;
+  aDiagonalChanged := gMapElements[Land[Y,X].Obj].DiagonalBlocked xor gMapElements[aObject].DiagonalBlocked;
 
+  Land[Y, X].Obj := aObject;
+  Result := True;
   //Apply change
   //UpdatePassability and UpdateWalkConnect are called in SetField so that we only use it in trees and other objects
   case aObject of
     88..124,
     126..172: // Trees - 125 is mushroom
               begin
-                Land[Y, X].Obj := aObject;
                 if ObjectIsChopableTree(KMPoint(X,Y), caAge1) then Land[Y,X].TreeAge := 1;
                 if ObjectIsChopableTree(KMPoint(X,Y), caAge2) then Land[Y,X].TreeAge := TREE_AGE_1;
                 if ObjectIsChopableTree(KMPoint(X,Y), caAge3) then Land[Y,X].TreeAge := TREE_AGE_2;
                 if ObjectIsChopableTree(KMPoint(X,Y), caAgeFull) then Land[Y,X].TreeAge := TREE_AGE_FULL;
-                UpdatePassability(KMRect(X, Y, X, Y)); //When using KMRect map bounds are checked by UpdatePassability
-                UpdateWalkConnect([wcWalk, wcRoad, wcWork], KMRectGrowTopLeft(KMRect(X, Y, X, Y)), DiagonalChanged);
-                Result := True;
               end
-    else      // Other objects
-              begin
-                Land[Y, X].Obj := aObject;
-                UpdatePassability(KMRect(X, Y, X, Y)); //When using KMRect map bounds are checked by UpdatePassability
-                UpdateWalkConnect([wcWalk, wcRoad, wcWork], KMRectGrowTopLeft(KMRect(X, Y, X, Y)), DiagonalChanged);
-                Result := True;
-              end;
   end;
+  if aUpdatePassability then
+  begin
+    UpdatePassability(KMRect(X, Y, X, Y)); //When using KMRect map bounds are checked by UpdatePassability
+    UpdateWalkConnect([wcWalk, wcRoad, wcWork], KMRectGrowTopLeft(KMRect(X, Y, X, Y)), aDiagonalChanged);
+  end;
+end;
+
+
+// Try to set an array of Tiles from script. Set terrain, rotation, height and object.
+// Update Passability, WalkConnect and Lighting only once at the end.
+// This is much faster, then set tile by tile with updates on every change
+//
+// Returns True if succeeded
+// use var for aTiles. aTiles can be huge so we do want to make its local copy. Saves a lot of memory
+function TKMTerrain.ScriptTrySetTilesArray(var aTiles: array of TKMTerrainTileBrief; aRevertOnFail: Boolean; var aErrors: TKMTerrainTileChangeErrorArray): Boolean;
+
+  function IsRectInit(aRect: TKMRect): Boolean;

```

Comment:
I added KMSameRect and KMRECT_INVALID, same as for Points. Then its better to call KMSameRect(Rect, KMRECT_INVALID)